usr/src/uts/common/fs/zfs/vdev.c

```diff
@@ -2489,7 +2489,12 @@ vdev_online(spa_t *spa, uint64_t guid, uint64_t flags, vdev_state_t *newstate)
 		spa_async_request(spa, SPA_ASYNC_CONFIG_UPDATE);
 	}
 
-	if (postevent)
+	if (wasoffline ||
+	    ((oldstate == VDEV_STATE_REMOVED ||
+	    oldstate == VDEV_STATE_CANT_OPEN ||
+	    oldstate == VDEV_STATE_FAULTED) &&
+	    (vd->vdev_state == VDEV_STATE_DEGRADED ||
+	    vd->vdev_state == VDEV_STATE_HEALTHY)))
```

Comment:
You will see this type of logic in other areas, so any change of order would require significant changes to the code. Feel free to expand the comment in zfs.h to make it more clear that order is significant.